Feature/isa data type

pkrog opened the pull request:
New data types ISA-tab and ISA-json, proposed by the [PhenoMeNal](https://github.com/phnmnl) team. These new types have been successfully used inside the PhenoMeNal project, in tools such as 
[mtbls-dwnld](https://github.com/workflow4metabolomics/mtbls-dwnld) and [isa2w4m](https://github.com/workflow4metabolomics/isa2w4m).

The uploader has been modified in order to be able to update the name of the composite dataset after uploading and opening the archive.

[isa-rwval](https://github.com/ISA-tools/isa-rwval) has been added as a new dependency. This package is required in order to read ISA archive content and display a summary to the user. It triggers the installation of packages `networkx` and `six`.

This PR is in relation with issue #3134 .

pkrog committed:
Add grooming for composite files.

Add call to groom_dataset_content() for composite files.

pkrog committed:
Add ISA composite data type.

Add new data types isa-tab and isa-json.
Add isa-rwval library (isatools Python 2 version) in requirements.

pcm32 committed:
[ISA datatype] adds logger setting for isatools.isatab to avoid failing uploads when no pandas (#31)

* [ISA datatype] adds logger setting for isatools.isatab to avoid failing uploads when no pandas.

* Update isa.py

pcm32 committed:
Installs isa-rwval from develop

Otherwise failing currently when building in the container.

pkrog committed:
Correct linting issues detected by Travis-CI run.

pkrog committed:
Correct again lint errors for Travis-CI.

pkrog committed:
Correct some more lint errors for Travis-CI.

pcm32 commented:
The following PR on starforge creates the wheel needed for the ISA Datatype, please review:
https://github.com/galaxyproject/starforge/pull/197

martenson commented on lib/galaxy/datatypes/isa.py:
```diff
@@ -0,0 +1,520 @@
+"""
+ISA datatype
+
+See http://isa-tools.org
```
This url seems down to me (and my colleague @natefoo), redirects to a non existing page at `http://apps1047.gold-traffic94.loan/?utm_medium=NQ3aDvyuBCtafRQJPeFC66tm%2bMNW8T%2baflxP0d0AJGo%3d&t=main4`

Not sure what is up with that.

pkrog replied:
The website address is correct. Except the site has been hacked. I will replace the URL with the GiHub URL for the organisation ISA tools.

martenson commented on lib/galaxy/dependencies/pinned-requirements.txt:
```diff
@@ -88,3 +88,7 @@ ecdsa==0.13
 python-genomespaceclient==0.1.8
+
+# ISA
+git+https://github.com/ISA-tools/isa-rwval.git@develop#egg=isa-rwval
```
Galaxy dependencies are handled using pipenv so in order to add deps we need to modify https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/dependencies/pipfiles/default/Pipfile and run the proper makefile target to rebuild this file - isntead of manual modification.

Also as mentioned on the starforge PR I think this should get to pypi first.

pkrog replied:
Is the wheel ready? May I replace this line with `isa-rwval==0.10.0` and test it?

pkrog committed:
Change ISA Tools web site link.

Replace ISA Tools .org web site that has just been hacked, by the GitHub
page of the organization.

martenson commented:
Unless there are compelling reasons for not doing so I am pushing this back to 18.09 since we most probably won't have time to properly test and review this PR for 18.05. 

@pkrog please let me know if this complicates things significantly for you or the PhenoMeNal team.

sneumann commented:
Hi, the pushback to 18.09 is a bit unfortunate, since our next release 18.08 will be based on Galaxy-18.05. Is there anything where @pkrog or @ilveroluca could help ?

pkrog commented:
Hi, if the wheel/pip package is ready, I think there's not much to do except updating requirement line. Am I wrong?

jmchilton commented on lib/galaxy/datatypes/isa.py:
```diff
@@ -0,0 +1,520 @@
+logger.handlers = []
+logger.propagate = False
+logger.addHandler(ch)
+logger.setLevel(logging.ERROR)
```
I sort of understand the logging above - does this logging need to be configured also?

pkrog replied:
Sorry @jmchilton , I fail to understand your remark. Could you elaborate?

jmchilton commented on lib/galaxy/datatypes/isa.py:
```diff
@@ -0,0 +1,520 @@
+                rval.append('<div>ISA Dataset composed of the following files:<p/><ul>')
+                for cmp_file in os.listdir(dataset.extra_files_path):
+                    opt_text = ''
+                    rval.append('<li><a href="%s" type="text/plain">%s</a>%s</li>' % (cmp_file, cmp_file, opt_text))
```
You escape above but not here, I think should escape these file names and such.

pkrog replied:
I'll escape the displayed filename, not the link, and remove `opt_text`, which is empty.

jmchilton commented on lib/galaxy/datatypes/isa.py:
```diff
@@ -0,0 +1,520 @@
+                        html += '<p>Data files:</p>'
+                        html += '<ul>'
+                        for data_file in assay.data_files:
+                            html += '<li>' + str(data_file.id) + ' - ' + str(data_file.filename) + ' - ' + str(data_file.label) + '</li>'
```
More escaping is needed in this method as well I believe.

jmchilton commented on lib/galaxy/datatypes/isa.py:
```diff
@@ -0,0 +1,520 @@
+        # extract archive if the file corresponds to the ISA archive
+        if basename == ISA_ARCHIVE_NAME:
+            # perform extraction
+            with open(file_name, 'rb') as stream:
```
Does 

```
from galaxy.util.compression_utils import CompressedFile
CompressedFile(file_name).extract(output_path)
```

work here? It'd simplify the implementation a lot I think if you did that. Could then get rid of _extract_* helpers, maybe that regex at the top, etc....

jmchilton commented on tools/data_source/upload.py:
```diff
@@ -241,9 +249,17 @@ def add_composite_file(dataset, output_path, files_path):
+                shutil.move(dp, file_output_path)
+                # groom the dataset file content if required by the corresponding datatype definition
+                if datatype.dataset_content_needs_grooming(file_output_path):
+                    datatype.groom_dataset_content(file_output_path)
```
Any chance I could get you to write a test case for this - we have really been working on increasing our test coverage for upload and datatype options and such over the last year. 

The file https://github.com/galaxyproject/galaxy/blob/dev/test/api/test_tools_upload.py has a lot of test cases, in particular there is a composite upload test case here https://github.com/galaxyproject/galaxy/blob/dev/test/api/test_tools_upload.py#L119 that might be a useful template - I'm not sure. If you don't have the will or time to implement such a test case can you place a small example isa archive in test-data and I could maybe take a stab at it. I guess this change is needed to ensure the archive is uncompressed as part of the upload - I'm not sure groom_dataset_content is the right place to do that but I'm not sure it is wrong either - so a test case really might help clarify and make sure we don't break anything if we need to refactor later.

jmchilton replied:
I don't know if this is relevant but if you want to set the name dynamically from https://github.com/workflow4metabolomics/mtbls-dwnld/blob/develop/mtbls-dwnld.xml - I think I could help with that. Rather than change the data type I'd recommend setting the name in a tool provided metadata file - I could give you some links for that. If the problem is really about upload though that isn't helpful.

pkrog replied:
I'm interested in writing the test. I'll have a look at your pointers.

The name of the dataset is set correctly within mtbls-dwnld. So, no, the issue is with the uploader.

pkrog replied:
Hi @jmchilton , I've added a small zip file for testing the uploader, and started to write a method `test_composite_datatype_isatab`. Howerver I haven't figured out yet how to submit the content of the zip file to the uploader (or the file path if this is sufficient). Could you help me please?

pkrog replied:
Hi @jmchilton , I see that our PR is still scheduled for the 18.09 release, however as stated by @sneumann it is really important for us that it gets included in the 18.05 release. The only issue remaining seems to be this uploader test. If you help me on this, I can finish it.

djcomlab replied:
@pkrog @jmchilton I've updated the test.

jmchilton commented on lib/galaxy/datatypes/isa.py:
```diff
@@ -0,0 +1,520 @@
+        if investigation is not None:
+            dataset.name = investigation.identifier
+        else:
+            dataset.name = 'ISA DATASET'
```
Would it be better to create some metadata on this datatype and set this there - I'm a bit worried about overriding the behavior of the upload datatype and the names it creates on a per-datatype basis. I do get this increases the usability of this datatype - but suddenly the upload API isn't going to work the way on would expect if you set a name explicitly or something - and we are constantly debating how to build default names from URL uploads - this would seem to override any decision on that as well. Even if you can make a case for using the investigation if set, it shouldn't be setting the default like this is the investigation is not set right?

pkrog replied:
The `else` part is certainly not useful.
I've noticed that the uploader set the dataset name to the file name (for instance `MTBLS354.zip`), now. I don't remember it was doing that before, but maybe I'm wrong. The filename should be sufficient for the user I guess.

pkrog committed:
Escape filename in primary file.

pkrog committed:
Escape assay information in ISA display_data().

pkrog committed:
Remove unused method _list_archive_files().

pkrog committed:
Use CompressedFile class.

Replace unzip custom code by Galaxy call to CompressedFile class.

pkrog committed:
Remove logger configuration.

djcomlab commented on lib/galaxy/datatypes/isa.py:
```diff
@@ -0,0 +1,395 @@
+                    html += '<p>Measurement type: %s</p>' % assay.measurement_type.term  # OntologyAnnotation
+                    html += '<p>Technology type: %s</p>' % assay.technology_type.term    # OntologyAnnotation
+                    html += '<p>Technology platform: %s</p>' % assay.technology_platform
+                    if assay.data_files is not None:
```
Note that this bit doesn't actually output any of the assay data files as it stands, because you're only using the investigation file parser from `isa-rwval`. So in all cases this data files list will be empty

pkrog replied:
Yes, I've noted that it's always empty now. That's unfortunate. Any workaround?

djcomlab replied:
I think this has become like this because the original request for `isa-rwval` was to support Python 2.7 (since full `isatools` is Python 3), but then it turns out that for Galaxy we don't want to add `pandas` to the requirements. The study and assay table parsing relies on `pandas` currently.

I can probably add to `isa-rwval` some features to extract the data files (and possibly a little bit more other useful metadata). Once I push a new version, the pinned-requirements.txt may need to be updated.

pkrog replied:
That would be great. But maybe not for this release, unless you have time to do it now. I'll comment these lines in the meantime.

djcomlab replied:
I can do it today.

pkrog replied:
Great job @djcomlab , I've just tested the modifications on my computer. It works well.

pkrog committed:
Comment out raw files listing in ISA type.

djcomlab commented on lib/galaxy/datatypes/isa.py:
```diff
@@ -329,12 +329,13 @@ def display_data(self, trans, dataset, preview=False, filename=None, to_ext=None
-                        for data_file in assay.data_files:
-                            html += '<li>' + escape(util.unicodify(str(data_file.id), 'utf-8')) + ' - ' + escape(util.unicodify(str(data_file.filename), 'utf-8')) + ' - ' + escape(util.unicodify(str(data_file.label), 'utf-8')) + '</li>'
-                        html += '</ul>'
+                    # The list of Raw data files contained in the assay file cannot be obtained anymore, due to limitation in isa-rwval library. David Johnson (@djcomlab) is working on a solution, and will update isa-rwval soon.
```
It's technically not a limitation of isa-rwval. It does it, but needs Pandas to do it. The limitation is because Galaxy does not want to add Pandas to its requirements 😁

pkrog committed:
Remove dynamical setting of study name for ISA.

machinelit committed:
Add new lazy table parsing for isa-tab datatype and reintroduce data files

pkrog committed:
Start writing upload test for isa-tab.

Do not know how to read zip file content and submit it to uploader for a
composite type.

machinelit committed:
Introduce new table parsers to get data files back; temporarily puts dependency back to Git repo to test against (updated PyPI package for isa-rwval will be pushed soon).

djcomlab commented on lib/galaxy/dependencies/pinned-requirements.txt:
```diff
@@ -91,4 +91,3 @@ python-genomespaceclient==0.1.8
 
 # ISA
 git+https://github.com/ISA-tools/isa-rwval.git@develop#egg=isa-rwval
```
Actually can now revert this line back to the package name.

`isa-rwval==0.10.1` should now allow some table parsing to reintroduce the data files displaying.

machinelit committed:
Fix requirements again for isa-rwval==0.10.1

machinelit committed:
Cleanup presentation of data files (unused rendering using datafile id)

pkrog committed:
Pinned requirements as a symbolic link as done in dev branch.

machinelit committed:
Add check so that empty filenames not added to file list.

machinelit committed:
Fix upload test for isa-tab datatype

machinelit committed:
Update isa-rwval version as new update fixes a bug

martenson commented:
@jmchilton can you please give this a review?

pkrog committed:
Merge branch 'dev' into feature/isa_data_type

fgiacomoni commented:
@martenson and @jmchilton, do you have a short moment to review this PR. 
I'm back on this thread because the isa datatype is central in our two galaxy public servers (Workflow4metabolomics and H2020 Phenomenal) dedicated in metabolomics. 
In which galaxy release (18.09?) do you plan to merge the PR? A big thanks

pkrog committed:
Remove spaces in blank line.

pkrog committed:
Various corrections for py34-lint

Add second blank line.
Remove useless imports.

martenson commented:
@galaxybot test this

martenson commented:
@pkrog @fgiacomoni @djcomlab This will be part of 19.01 Galaxy release. Thank you for the contribution and sorry for the delay. 

We might get back to you asking for some tests I think, but I don't expect any big problems. Thanks again. 👍

ping @jmchilton

fgiacomoni commented:
Great - @martenson thanks for the review (thanks to @bgruening for the ping 😄  ) and thanks @pkrog and @djcomlab and @jmchilton for isa effort integration. @proccaserra should be interested by this feature!

pkrog commented:
Many thanks @martenson , @bgruening and @jmchilton .
And we should not forget @kikkomep for his great work too in this PR, thanks Marco !
